### ISSUE TITLE
Vector contents can move without moving the vector itself

### ISSUE THREAD

**Issue by SimonSapin** [Manishearth/rust-tenacious#1]
From the README:
> Note that #[no_move] is transitive, any struct or enum containing a #[no_move] type must be annotated as well. Similarly, any type with #[no_move] substitutions in its type parameters (E.g. Vec<Foo> where Foo is no_move) will be treated as immovable.

That’s an interesting example. I assume “not moving” means “staying at the same address in memory so that raw pointers to it stay valid”?

When a vector is at capacity, `Vec::push` reallocates a larger chunk of memory and moves its contents there. So `Foo` can move even if `Vec<Foo>` doesn’t move. Does this lint cover this kind of case as well?

**Comment by Manishearth** [comment]
No, it doesn't. We'll have to figure out a way to lint on Vec methods specifically.

**Comment by SimonSapin** [comment]
My point is not just about `Vec`. It’s that there are many causes of moves, and trying to patch them one by one does not sound like a good approach.

**Comment by Manishearth** [comment]
Well, this is a lint -- it can't handle *all* situations..

What I can do is:

 - warn on rvalue moves (which I plan to do anyway) except for `let foo = gen_rvalue()`
 - add an extra note to moves into functions that are defined with a generic parameter.

**Comment by Manishearth** [comment]
Blocked on https://github.com/rust-lang/rust/issues/25005
